google/cloud/security/scanner/scanners/groups_scanner.py

```diff
@@ -50,66 +52,114 @@ def __init__(self, global_configs, snapshot_timestamp):
 
         Args:
-            timestamp: String of snapshot timestamp, formatted as
-                YYYYMMDDTHHMMSSZ.
+            root: The nodes (tree structure) to find violations in.
 
         Returns:
-            The root node that holds the tree structure of all the groups
-                in the organization.
+            A list of nodes that are in violation.
         """
-        root = MemberNode(MY_CUSTOMER, MY_CUSTOMER)
+        all_violations = []
+        for node in anytree.iterators.PreOrderIter(root):
 
-        all_groups = self.dao.get_all_groups('groups', timestamp)
-        for group in all_groups:
-            group_node = MemberNode(group.get('group_id'),
-                                    group.get('group_email'),
-                                    'group',
-                                    'ACTIVE',
-                                    root)
-            group_node = self.get_recursive_members(group_node, timestamp)
+            # No need to evaluate these nodes.
+            # This represents the org, i.e. is not a group.
+            if node.member_email == MY_CUSTOMER:
+                continue
+            # This represents the auto-generated group, containing all the users
+            # in the org.
+            if not node.member_email:
+                continue
 
-        LOGGER.debug(anytree.RenderTree(
-            root, style=anytree.AsciiStyle()).by_attr('member_email'))
+            node.violated_rule_names = []
+            whitelist_rule_statuses = []
+            for rule in node.rules:
+                condition_statuses = []
 
-        return root
+                if rule.get('mode') == 'whitelist':
+                    for condition in rule.get('conditions'):
+                        if condition.get('member_email') in node.member_email:
+                            condition_statuses.append(True)
+                        else:
+                            condition_statuses.append(False)
+
+                    # All the conditions of this rule have evaluated.
+                    # The rule is fulfilled, if any condition matches.
+                    if any(condition_statuses):
+                        whitelist_rule_statuses.append(True)
+                    else:
+                        whitelist_rule_statuses.append(False)
+                        node.violated_rule_names.append(rule.get('name'))
+
+                elif rule.get('mode') == 'blacklist':
+                    pass  # TODO
+
+                elif rule.get('mode') == 'required':
+                    pass  # TODO
+
+                else:
+                    pass  # TODO
+
+            # Determine if the node is in violations or not.
+            # All rules must be fulfilled, for a node to not be in violation.
+            # If any rule is not fulfilled, then node is in violation.
+            #
+            # truth table
+            # http://stackoverflow.com/a/19389957/2830207
```

Comment:
not necessary to put this here, people who don't know how any() works can google it